clients/include/testing_spgemmreuse_csr.hpp

```diff
@@ -0,0 +1,600 @@
+                                                   descr,
+                                                   &bufferSize2,
+                                                   externalBuffer2,
+                                                   &bufferSize3,
+                                                   externalBuffer3,
+                                                   &bufferSize4,
+                                                   externalBuffer4));
+
+    CHECK_HIP_ERROR(hipMalloc(&externalBuffer2, bufferSize2));
+    CHECK_HIP_ERROR(hipMalloc(&externalBuffer3, bufferSize3));
+    CHECK_HIP_ERROR(hipMalloc(&externalBuffer4, bufferSize4));
+
+    CHECK_HIPSPARSE_ERROR(hipsparseSpGEMMreuse_nnz(handle,
+                                                   transA,
+                                                   transB,
+                                                   A,
+                                                   B,
+                                                   C1,
+                                                   alg,
+                                                   descr,
+                                                   &bufferSize2,
+                                                   externalBuffer2,
+                                                   &bufferSize3,
+                                                   externalBuffer3,
+                                                   &bufferSize4,
+                                                   externalBuffer4));
+
+    // We can already free buffer1
+    CHECK_HIP_ERROR(hipFree(externalBuffer1));
+    CHECK_HIP_ERROR(hipFree(externalBuffer2));
+
+    // Get nnz of C
+    int64_t rows_C, cols_C, nnz_C_1, nnz_C_2;
+    CHECK_HIPSPARSE_ERROR(hipsparseSetPointerMode(handle, HIPSPARSE_POINTER_MODE_HOST));
+    CHECK_HIPSPARSE_ERROR(hipsparseSpMatGetSize(C1, &rows_C, &cols_C, &nnz_C_1));
+    nnz_C_2 = nnz_C_1;
+
+    // Allocate C
+    auto dcsr_col_ind_C_1_managed
+        = hipsparse_unique_ptr{device_malloc(sizeof(J) * nnz_C_1), device_free};
+    auto dcsr_val_C_1_managed
+        = hipsparse_unique_ptr{device_malloc(sizeof(T) * nnz_C_1), device_free};
+    auto dcsr_col_ind_C_2_managed
+        = hipsparse_unique_ptr{device_malloc(sizeof(J) * nnz_C_2), device_free};
+    auto dcsr_val_C_2_managed
+        = hipsparse_unique_ptr{device_malloc(sizeof(T) * nnz_C_2), device_free};
+
+    J* dcsr_col_ind_C_1 = (J*)dcsr_col_ind_C_1_managed.get();
+    T* dcsr_val_C_1     = (T*)dcsr_val_C_1_managed.get();
+    J* dcsr_col_ind_C_2 = (J*)dcsr_col_ind_C_2_managed.get();
+    T* dcsr_val_C_2     = (T*)dcsr_val_C_2_managed.get();
+    if(!dcsr_col_ind_C_1 || !dcsr_val_C_1 || !dcsr_col_ind_C_2 || !dcsr_val_C_2)
+    {
+        verify_hipsparse_status_success(
+            HIPSPARSE_STATUS_ALLOC_FAILED,
+            "!dcsr_col_ind_C_1 || !dcsr_val_C_1 || !dcsr_col_ind_C_2 || !dcsr_val_C_2");
+
+        return HIPSPARSE_STATUS_ALLOC_FAILED;
```

Comment:
Note if we return here then we are not free'ing all of your hipMalloc's (specifically buffer 3 and 4)

Reply:
good catch, I need to wrap these buffers. any hip return would leave these unfreed anyway.

Reply:
wrapped it with unique ptr and reset when needed.